package ObserverDatagramPkg;

// Chamada do compilador do Protocol Buffer
// ..\..\..\..\dependencies\protobuf\Release\protoc.exe --cpp_out=. protocol.proto

enum TypesOfSubjects 
{
    TObsUnknown         =  0;

    TObsCell            =  1;
    TObsCellularSpace   =  2;
    TObsNeighborhood    =  3;
    TObsTimer           =  4;
    TObsEvent           =  5;
    TObsTrajectory      =  6;
    TObsAutomaton       =  7;
    TObsAgent           =  8; 
    TObsEnvironment     =  9;
    TObsSociety         = 10; 

    // TObsMessage,            // it isn't a Subject
    // TObsState,              // it isn't a Subject
    // TObsJumpCondition,      // it isn't a Subject
    // TObsFlowCondition,      // it isn't a Subject
}

enum TypesOfObservers
{
    TObsUndefined       =  0;

    TObsTextScreen      =  1;
    TObsLogFile         =  2;
    TObsTable           =  3;
    TObsGraphic         =  4;
    TObsDynamicGraphic  =  5;
    TObsMap             =  6;
    TObsUDPSender       =  7;
    TObsScheduler       =  8;
    TObsImage           =  9;
    TObsStateMachine    = 10;
	TObsNeigh			= 11;
    TObsShapefile       = 12;
    TObsTCPSender       = 13;
}

// enum TypesOfData
//{
//    TObsBool            =   0;
//    TObsNumber          =   1;
//    TObsDateTime        =   2;
//    TObsText            =   4;
//
//    TObsUnknownData     = 100;
//}

// /////////////////////////////////////////////////////
// /////////////////////////////////////////////////////

message RawAttribute
{
    // required TypesOfData type = 1;
    // required string key = 2;
    
    // optional double number = 3;
    // optional string text = 4;
    
    required string key = 1;
    
    optional double number = 2;
    optional string text = 3;
}
        
message SubjectAttribute
{
    required int32 id = 1;
    required TypesOfSubjects type = 2;
    required int32 attribsNumber = 3;
    required int32 itemsNumber = 4;
    
    repeated RawAttribute rawAttributes = 5;
    repeated SubjectAttribute internalSubject = 6;
}

